### PR TITLE
Closes #712: Count whole dataset download through API

### LINKED ISSUE #712: Explore if is it possible to add download whole dataset log in ui without major performance issues

### PR DESCRIPTION
**Pull request by wtr-icm**

### CODE REVIEW

File: dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/api/Access.java [hunk 1]
```diff
@@ -152,6 +157,8 @@
     @Inject
     private EmbargoAccessService embargoAccessService;
 
+    @Inject
+    private WholeDatasetDownloadLogger wholeDatasetDownloadLogger;
```
Comment (dbojanek) [inline dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/api/Access.java:160]: To consider: I'd rename it and drop `whole` from the name. We don't have a non-whole download logger so it doesn't really tell us much.

File: dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/api/Access.java [hunk 2]
```diff
@@ -493,177 +495,154 @@ public DownloadInstance tabularDatafileMetadataPreprocessed(@PathParam("fileId")
...
-
+        final long zipDownloadSizeLimit = determineDownloadSizeLimit();
         logger.fine("setting zip download size limit to " + zipDownloadSizeLimit + " bytes.");
 
-        if (fileIds == null || fileIds.equals("")) {
-            throw new BadRequestException();
-        }
-
-        String apiToken = (apiTokenParam == null || apiTokenParam.equals(""))
+        String apiToken = (apiTokenParam == null || StringUtils.EMPTY.equals(apiTokenParam))
```
Comment (dbojanek) [inline dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/api/Access.java]: We tend to replace logic like that. Consider using `StringUtils.isEmpty(apiTokenParam)`

File: dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/api/Access.java [hunk 3]
```diff
@@ -493,177 +495,154 @@ public DownloadInstance tabularDatafileMetadataPreprocessed(@PathParam("fileId")
...
-                            if (file != null) {
-                                if (isAccessAuthorized(file, apiToken)) {
-
-                                    logger.fine("adding datafile (id=" + file.getId() + ") to the download list of the ZippedDownloadInstance.");
-                                    //downloadInstance.addDataFile(file);
-                                    if (gbrecs == null && file.isReleased()) {
-                                        GuestbookResponse gbr = guestbookResponseService.initAPIGuestbookResponse(file.getOwner(), file, session, apiTokenUser);
-                                        guestbookResponseService.save(gbr);
-                                    }
+        final boolean getOriginal = shouldSendOriginalFormat(uriInfo.getQueryParameters());
```
Comment (dbojanek) [inline dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/api/Access.java]: Boolean variables naming convention infer using prefixes like `isOriginal`, `hasOrigina` and such. `get` is for getters

File: dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/api/Access.java [hunk 4]
```diff
@@ -493,177 +495,154 @@ public DownloadInstance tabularDatafileMetadataPreprocessed(@PathParam("fileId")
...
+
+    private long determineDownloadSizeLimit() {
+        long limit = settingsService.getValueForKeyAsLong(SettingsServiceBean.Key.ZipDownloadLimit);
+        if (limit == -1) {
+            throw new BadRequestException("Download zipped bundles of multiple files option is disabled in this installation");
+        }
+        return limit != 0 ? limit : Long.MAX_VALUE;
+    }
+
+    private void assertOrThrow(Supplier<Boolean> constraint) {
```
Comment (dbojanek) [inline dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/api/Access.java]: I'd either change the name to indicate what is being thrown eg. `assertOrBadRequest`, `assertTrueOrBadRequest` ect. or generalize method and add parameter with exception being thrown.

File: dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/api/Access.java [hunk 5]
```diff
@@ -493,177 +495,154 @@ public DownloadInstance tabularDatafileMetadataPreprocessed(@PathParam("fileId")
...
+        if (!constraint.get()) {
+            throw new BadRequestException();
+        }
+    }
+
+    private boolean shouldSendOriginalFormat(MultivaluedMap<String, String> queryParameters) {
```
Comment (dbojanek) [inline dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/api/Access.java]: For boolean methods naming convention is to use verbs like `is` `are` `was` ect. So I'd refactor it to something like `isOriginalFormatSend`

File: dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/datafile/page/WholeDatasetDownloadLogger.java [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+
+    @Inject
+    public WholeDatasetDownloadLogger(DownloadDatasetLogService downloadDatasetLogService) {
+        this.downloadDatasetLogService = downloadDatasetLogService;
+    }
+
+    // -------------------- LOGIC --------------------
+
+    public void incrementLogIfDownloadingWholeDataset(List<DataFile> filesToDownload) {
+        if (filesToDownload == null || filesToDownload.isEmpty()) {
```
Comment (dbojanek) [inline dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/datafile/page/WholeDatasetDownloadLogger.java]: `Lists.isEmpty()` ?

**Review by dbojanek — CHANGES_REQUESTED** [review]

File: dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/api/Access.java (hunk 5 above)
Comment (wtr-icm) [reply dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/api/Access.java]: I've changed to better name, but personally I think that ```should``` for boolean-returning method is also ok ;)

File: dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/api/Access.java (hunk 1 above)
Comment (wtr-icm) [reply dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/api/Access.java:160]: I see this differently: when we have simple ```DatasetDownload``` we don't pay attention whether it concerns downloading dataset as a whole or some part of it. When, on the other hand, we have ```WholeDatasetDownload``` we know immediately that only whole dataset download is important.

File: dataverse-persistence/src/main/resources/db/migration/V21__recreateDownloadDatasetLogTable.sql [hunk 7]
```diff
@@ -0,0 +1,9 @@
+DROP TABLE IF EXISTS downloaddatasetlog;
+CREATE TABLE downloaddatasetlog (
+	id bigint PRIMARY KEY,
+	dataset_id bigint REFERENCES dataset NOT NULL,
+	downloaddate timestamp NOT NULL
+);
+
+CREATE INDEX downloaddatasetlog_dataset_id_idx ON downloaddatasetlog (dataset_id);
+CREATE SEQUENCE downloaddatasetlog_id_seq AS bigint INCREMENT BY 50 MINVALUE 50;
```
Comment (dbojanek) [inline dataverse-persistence/src/main/resources/db/migration/V21__recreateDownloadDatasetLogTable.sql:9]: Why would we want ids to start from 50 and increment by 50? This ought to be a simple id column.

File: dataverse-persistence/src/main/java/edu/harvard/iq/dataverse/persistence/dataset/DownloadDatasetLog.java [hunk 8]
```diff
@@ -2,30 +2,52 @@
...
+import javax.persistence.TemporalType;
+import java.util.Date;
 
 @Entity
 public class DownloadDatasetLog {
     @Id
+    @GeneratedValue(strategy = GenerationType.SEQUENCE, generator = "downloaddatasetlog_id_seq_gen")
+    @SequenceGenerator(name = "downloaddatasetlog_id_seq_gen", sequenceName = "downloaddatasetlog_id_seq", allocationSize = 50)
+    @Column(name = "id")
+    private Long id;
```
Comment (dbojanek) [inline dataverse-persistence/src/main/java/edu/harvard/iq/dataverse/persistence/dataset/DownloadDatasetLog.java:19]: Everywhere in the app we use just:
`    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;`

At least for consistency IMO we should stick to it.

File: dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/api/Access.java (hunk 1 above)
Comment (dbojanek) [reply dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/api/Access.java:160]: I won't press the matter and your point is valid. I just see the logic behind counting dataset downloads as implementation detail that shouldn't affect how we see business rules that we are trying to meet here.

File: dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/datafile/page/WholeDatasetDownloadLogger.java [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+
+        logger.log(Level.FINE, "Checking for whole dataset download in datasets [{0}] took: {1} ms",
+                new Object[] { extractDatasetIds(filesByDataset), totalTime });
+    }
+
+    // -------------------- PRIVATE --------------------
+
+    private Map<Dataset, List<DataFile>> groupFilesByDatasetsAndUpdateLogIfNeeded(List<DataFile> filesToDownload) {
+        Map<Dataset, List<DataFile>> filesByDataset = filesToDownload.stream()
+                .filter(f -> f != null && f.getOwner() != null)
```
Comment (dbojanek) [inline dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/datafile/page/WholeDatasetDownloadLogger.java:60]: What does it mean for a `DataFile` to not have owner? Is this possible?

File: dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/datafile/page/WholeDatasetDownloadLogger.java [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+
+    private Set<Long> extractIdsOfFiles(List<DataFile> dataFiles) {
+        return dataFiles.stream()
+                .map(DataFile::getId)
+                .collect(toSet());
+    }
+
+    private List<Set<Long>> createPossibleSetsOfFileIdsFromVersions(Dataset dataset, Set<Long> fileIdSet) {
+        return Optional.ofNullable(dataset)
+                .filter(d -> !d.hasActiveEmbargo())
```
Comment (dbojanek) [inline dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/datafile/page/WholeDatasetDownloadLogger.java]: I don't think we should consider embargo like that. 
First of all, it's still possible for specific users to download files from a dataset under embargo. 
Secondly embargo is on a dataset level so the way it works now is `if there is embargo filter out all versions and handle empty list` which is not intuitive.

File: dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/datafile/page/WholeDatasetDownloadLogger.java [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+            downloadDatasetLogService.logWholeSetDownload(dataset.getId());
+        }
+    }
+
+    /**
+     * Here we are extracting sets of ids of files from all the versions that have no more files than
+     * list of files being downloaded and we check if any of these sets is wholly contained in the set of ids
+     * of files being downloaded
+     */
+    private boolean isUserDownloadingAllFilesFromAnyVersion(Dataset dataset, List<DataFile> dataFiles) {
```
Comment (dbojanek) [inline dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/datafile/page/WholeDatasetDownloadLogger.java:79]: In tandem with comment on `createPossibleSetsOfFileIdsFromVersions`.

The way I'd imagine it work:
1. Check embargo on dataset (if there is one should we consider users permissions or just abort? - maybe discuss with @wfenrich )
2. Generate sets of filesids 
3. Find out if `wholedatasetdownload` is happening

**Review by dbojanek — CHANGES_REQUESTED** [review]

File: dataverse-persistence/src/main/resources/db/migration/V21__recreateDownloadDatasetLogTable.sql (hunk 7 above)
Comment (wtr-icm) [reply dataverse-persistence/src/main/resources/db/migration/V21__recreateDownloadDatasetLogTable.sql:9]: It does not work that way: we have EclipseLink as JPA provider and for that reason the sequence starting point is chosen to that value in order to start numbering from 1 (it does something like: Seq_start_value - Sequence_increment + 1, 2, 3…).
Also – ids will be numbered continuously, but we will have much less database queries (only one for 50 inserts, while with IDENTITY we have SELECT after every insert).

File: dataverse-persistence/src/main/java/edu/harvard/iq/dataverse/persistence/dataset/DownloadDatasetLog.java (hunk 8 above)
Comment (wtr-icm) [reply dataverse-persistence/src/main/java/edu/harvard/iq/dataverse/persistence/dataset/DownloadDatasetLog.java:19]: Yes, you're right, but I think that there are cases, when more efficient solutions should prevail over code consistency. In this case we have already many queries going to db (querying for datasets, file metadata, versions etc.), so when we can reduce their total number, we should do it.

File: dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/datafile/page/WholeDatasetDownloadLogger.java (hunk 9 above)
Comment (wtr-icm) [reply dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/datafile/page/WholeDatasetDownloadLogger.java:60]: I do not know. However I prefer to be prepared for such cases.

File: dataverse-persistence/src/main/java/edu/harvard/iq/dataverse/persistence/dataset/DownloadDatasetLog.java (hunk 8 above)
Comment (dbojanek) [reply dataverse-persistence/src/main/java/edu/harvard/iq/dataverse/persistence/dataset/DownloadDatasetLog.java:19]: You're right. But my intuition coming from big data background tells me that database will be idle most of the time and we are not harnessing its capabilities. I guess we lack stress analysis and forecasts on system usage.

**Review by dbojanek — APPROVED** [review]

File: dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/datafile/page/WholeDatasetDownloadLogger.java (hunk 11 above)
Comment (wtr-icm) [reply dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/datafile/page/WholeDatasetDownloadLogger.java:79]: If I correctly understand, after removing check for embargo, this should behave as you've written.

Comment (dbojanek) [reply dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/datafile/page/WholeDatasetDownloadLogger.java:79]: Yes, we have agreed on removing embargo checks while counting downloads. If downloads happens we don't bother checking for possible reasons why it did, we just count it.

**Review by dbojanek — APPROVED** [review]

**Pull request merged**